### ISSUE TITLE
Incompatible Storage API Proxy

### ISSUE THREAD

**Issue by ibnesayeed** [webrecorder/wombat#65]
The following Proxy overrides the Storage API in a way that is not compatible with the native implementations, causing issues in some archival replays.

https://github.com/webrecorder/wombat/blob/4f91f898504caca53dd18e8232dd04cedbc41256/src/wombat.js#L5760-L5775

Below is an overview of some of the incompatibilities.

## Underlying object structure

```js
/* Native */
localStorage
//=> Storage {length: 0}
typeof localStorage
//=> "object"
Object.keys(localStorage)
//=> []
Object.getOwnPropertyNames(localStorage)
//=> []
```

```js
/* Wombat */
localStorage
//=> Proxy {data: {…}, wombat: Wombat, proxying: "localStorage", _deleteItem: ƒ}
typeof localStorage
//=> "object"
Object.keys(localStorage)
//=> []
Object.getOwnPropertyNames(localStorage)
//=> ["data", "wombat", "proxying", "_deleteItem"]
```

It uses common terms like `data` as keys in the implementation, which have higher chances of collisions (consequences described later).

## Missing key access

```js
/* Native */
localStorage.getItem('foo')
//=> null
localStorage.foo
//=> undefined
localStorage['foo']
//=> undefined
```

```js
/* Wombat */
localStorage.getItem('foo')
//=> null
localStorage.foo
//=> null
localStorage['foo']
//=> null
```

The native implementation returns `null` value for missing keys when using `getItem()`, but `undefined` when using attribute accessor (i.e., dot notation or bracket notation). However, Wombat implementation returns `null` in all the cases. We encountered a site that uses `localStorage.foo !== undefined` as a condition, which breaks when replayed with Wombat.

## Storing data

```js
/* Native */
localStorage.setItem('foo', 'FOO')
//=> undefined
localStorage.getItem('foo')
//=> "FOO"
localStorage.foo
//=> "FOO"
localStorage['foo']
//=> "FOO"
Object.keys(localStorage)
//=> ["foo"]
Object.getOwnPropertyNames(localStorage)
//=> ["foo"]
```

```js
/* Wombat */
localStorage.setItem('foo', 'FOO')
//=> undefined
localStorage.getItem('foo')
//=> "FOO"
localStorage.foo
//=> "FOO"
localStorage['foo']
//=> "FOO"
Object.keys(localStorage)
//=> []
Object.getOwnPropertyNames(localStorage)
//=> ["data", "wombat", "proxying", "_deleteItem"]
```

Storing data works both using the `setItem()` and attribute accessor (the latter is not illustrated, but tested). However, stored keys are not exposed the way they are done in the native implementation. This means, membership checking may fail in some web pages.

## Storing reserved keys

```js
/* Native */
localStorage.data = 'DATA'
//=> "DATA"
localStorage.data
//=> "DATA"
localStorage.getItem('data')
//=> "DATA"
```

```js
/* Wombat */
localStorage.data = 'DATA'
//=> "DATA"
localStorage.data
//=> {foo: "FOO"}
localStorage.getItem('data')
//=> null
```

When a is stored using a reserved key (in this case, `data`) using the attribute accessor style (i.e., dot or bracket notations), it is silently ignored. However, this restriction can be bypassed when using the `setItem()` method, as show below:

```js
/* Wombat */
localStorage.setItem('data', 'DATA')
//=> undefined
localStorage.getItem('data')
//=> "DATA"
localStorage.data
//=> {foo: "FOO", data: "DATA"}
```

However, this can only be read reliably using the `getItem()` method, because when using the attribute accessor style, it returns the encapsulating object.

**Comment by ibnesayeed** [comment]
Here is an example memento that is affected by this bug.

https://web.archive.org/web/20210628064337/http://scripting.com/

The block of code from lines 1114-1117 (shown below) of https://web.archive.org/web/20210702034024js_/http://scripting.com/code/oldschool/daytemplate.js is triggering a JS error where `savedState` is `null` because `localStorage.savedState` is not set. On the live site this block is not executed initially due to the `undefined` check, but Wombat Storage prosy returns `null` for `localStorage.savedState` instead of `undefined`.

```js
	if (localStorage.savedState !== undefined) {
		savedState = JSON.parse (localStorage.savedState);
		savedState.currentTab = "blog"; //4/6/20 AM by DW
		}
```

**Comment by ibnesayeed** [comment]
Added test cases in #66 to address this issue.

**Comment by ikreymer** [comment]
Fixed via #68!
